In [49]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = "C:/Users/hardi/Documents/Instacart_Baskey_analysis"

In [3]:
#import customers dataset
cus = pd.read_csv(os.path.join(path, "Data","original_data","customers.csv"))

In [4]:
cus.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [7]:
cus.shape

(206209, 10)

In [21]:
#renaming columns
cus = cus.rename(columns={"Surnam":"last_name","STATE": "State","n_dependants":"Number_dependants", "fam_status":"Family_status"})

In [22]:
cus.head()

user_id First Name last_name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number_dependants Family_status  income  
0                  3       married  165665  
1                  0        single   59285  
2                  2       married   99568  
3                  0        single   42049  
4                  1       married   40374

In [23]:
#checking of dataset has any null values
cus.isnull().sum()

user_id                  0
First Name           11259
last_name                0
Gender                   0
State                    0
Age                      0
date_joined              0
Number_dependants        0
Family_status            0
income                   0
dtype: int64

In [24]:
#checking of dataset id any duplicated values
duplicates = cus.duplicated()

In [25]:
duplicates.sum()

0

In [26]:
#checking overview of dataset
cus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   user_id            206209 non-null  int64 
 1   First Name         194950 non-null  object
 2   last_name          206209 non-null  object
 3   Gender             206209 non-null  object
 4   State              206209 non-null  object
 5   Age                206209 non-null  int64 
 6   date_joined        206209 non-null  object
 7   Number_dependants  206209 non-null  int64 
 8   Family_status      206209 non-null  object
 9   income             206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


#### there are 11259 missing value in column first name. i won't drop row because i will losse information about customers. so i will replace with unknow

In [51]:
#replace null value in first name column with unknow
cus['First Name'].fillna('Unknown', inplace=True)

C:\Users\hardi\AppData\Local\Temp\ipykernel_21884\250771707.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  cus['First Name'].fillna('Unknown', inplace=True)


In [31]:
#checking to make sure ther is no missing value
cus.isnull().sum()

user_id              0
First Name           0
last_name            0
Gender               0
State                0
Age                  0
date_joined          0
Number_dependants    0
Family_status        0
income               0
dtype: int64

In [37]:
#check for mix data

for i in cus.columns.to_list():
    werid = (cus[i].map(type) != cus[i].iloc[0].__class__).any()
    if werid:
        print(i)

user_id
Age
Number_dependants
income


## combine dataset

In [42]:
ord_prod = pd.read_pickle(os.path.join(path,"Data", "prepare_data", "odrer_product_add_columns_combine_data_4.8.pkl"))

In [43]:
ord_prod.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   3139998      138            28                       6                 11   
1   1977647      138            30                       6                 17   
2    389851      709             2                       0                 21   
3    652770      764             1                       3                 13   
4   1813452      764             3                       4                 17   

   ...  _merge      busiest_day  busiest_days busiest_period_of_day  \
0  ...    both  regularbusy day   regular day           most orders   
1  ...    both  regularbusy day   regular day        Average orders   
2  ...    both      Busiest day   Busiest day        Average orders   
3  ...    both  regularbusy day   slowest day           most orders   
4  ...    both      Slowest Day   slowest day        Average orders   

  max_number_order      loyalty_flag mean_price_spent_customer  spending_flag  \
0               32  regular_customer                  6.935811  Lower spender   
1               32  regular_customer                  6.935811  Lower spender   
2                5      new_customer                  7.930208  Lower spender   
3                3      new_customer                  4.972414  Lower spender   
4                3      new_customer                  4.972414  Lower spender   

  customer_frequency     frequency_flag  
0                8.0  Frequent customer  
1                8.0  Frequent customer  
2                6.0  Frequent customer  
3                9.0  Frequent customer  
4                9.0  Frequent customer  

[5 rows x 23 columns]

In [44]:
#combine both dataset on column name user_id
merge_df = ord_prod.merge(cus,on='user_id',how='inner')

In [45]:
merge_df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   3139998      138            28                       6                 11   
1   1977647      138            30                       6                 17   
2    389851      709             2                       0                 21   
3    652770      764             1                       3                 13   
4   1813452      764             3                       4                 17   

   ...     frequency_flag  First Name  last_name  Gender      State Age  \
0  ...  Frequent customer     Charles        Cox    Male  Minnesota  81   
1  ...  Frequent customer     Charles        Cox    Male  Minnesota  81   
2  ...  Frequent customer     Deborah      Glass  Female    Vermont  66   
3  ...  Frequent customer     Heather      Myers  Female  Wisconsin  40   
4  ...  Frequent customer     Heather      Myers  Female  Wisconsin  40   

  date_joined  Number_dependants Family_status  income  
0    8/1/2019                  1       married   49620  
1    8/1/2019                  1       married   49620  
2   6/16/2018                  2       married  158302  
3    2/9/2020                  3       married   31308  
4    2/9/2020                  3       married   31308  

[5 rows x 32 columns]

In [46]:
merge_df.shape

(32404859, 32)

In [47]:
#check for mssing value in merge dataset
merge_df.isnull().sum()

product_id                   0
product_name                 0
aisle_id                     0
department_id                0
prices                       0
order_id                     0
user_id                      0
order_number                 0
orders_day_of_the_week       0
order_hour_of_day            0
days_since_last_order        0
add_to_cart_order            0
reordered                    0
_merge                       0
busiest_day                  0
busiest_days                 0
busiest_period_of_day        0
max_number_order             0
loyalty_flag                 0
mean_price_spent_customer    0
spending_flag                0
customer_frequency           0
frequency_flag               0
First Name                   0
last_name                    0
Gender                       0
State                        0
Age                          0
date_joined                  0
Number_dependants            0
Family_status                0
income                       0
dtype: i

In [48]:
#export merge dataset
merge_df.to_pickle(os.path.join(path,"Data", "prepare_data","customer_order_product_combine.pkl"))